internal/locate/region_cache.go

```diff
@@ -1203,6 +1204,149 @@ func (c *RegionCache) LocateKeyRange(bo *retry.Backoffer, startKey, endKey []byt
+		res = append(res, &KeyLocation{
+			Region:   r.VerID(),
+			StartKey: r.StartKey(),
+			EndKey:   r.EndKey(),
+			Buckets:  r.getStore().buckets,
+		})
+	}
+
+	// 2. load remaining regions from pd client
+	cachedIdx := 0
+	for len(uncachedRanges) > 0 {
+		regions, err := c.BatchLoadRegionsWithKeyRanges(bo, uncachedRanges, defaultRegionsPerBatch, opts...)
+		if err != nil {
+			return nil, err
+		}
+		if len(regions) == 0 {
+			return nil, errors.Errorf("BatchLoadRegionsWithKeyRanges return empty batchedRegions without err")
+		}
+
+		for uncachedIdx := 0; uncachedIdx < len(regions); {
+			if cachedIdx >= len(cachedRegions) {
+				for _, r := range regions[uncachedIdx:] {
+					appendRegion(r)
+				}
+				break
+			}
+			cachedRegion, uncachedRegion := cachedRegions[cachedIdx], regions[uncachedIdx]
+			var r *Region
+			if len(uncachedRegion.StartKey()) == 0 {
+				r = uncachedRegion
+				uncachedIdx++
+			} else if len(cachedRegion.StartKey()) == 0 {
+				r = cachedRegion
+				cachedIdx++
+			} else {
+				cmp := bytes.Compare(cachedRegion.StartKey(), uncachedRegion.StartKey())
+				switch cmp {
+				case 0, 1:
+					// loaded region: |<----
+					// cached region:  |<----
+					r = uncachedRegion
+					uncachedIdx++
+					if uncachedRegion.Contains(cachedRegion.EndKey()) {
+						// loaded region: |<------>|
+						// cached region:  |<---->|
+						// if the cached region is covered by the latest loaded region, the outdated cached region can be skipped.
+						cachedIdx++
+					}
+				case -1:
+					// loaded region:  |<----
+					// cached region: |<----
+					r = cachedRegion
+					cachedIdx++
+				}
+			}
+			appendRegion(r)
+		}
+
+		// if the regions are not loaded completely, split uncachedRanges and load the rest.
+		if lastRegion := regions[len(regions)-1]; len(lastRegion.EndKey()) == 0 || bytes.Compare(lastRegion.EndKey(), uncachedRanges[len(uncachedRanges)-1].EndKey) >= 0 {
+			// fast check, if all ranges are loaded from PD, quit the loop.
+			break
+		} else {
+			// binary search the first range that is not fully loaded from PD.
+			endKey := lastRegion.EndKey()
+			lower, upper := 0, len(uncachedRanges)-1
+			var mid int
+		SEARCH:
+			for lower <= upper {
+				mid = (lower + upper) / 2
+				switch bytes.Compare(endKey, uncachedRanges[mid].EndKey) {
+				case 0:
+					// region: ---->|
+					// ranges: ---->| |<---->|
+					// the current region exactly matches the range, just move the mid index to next region.
+					mid += 1
+					break SEARCH
+				case 1:
+					// region: ----->|
+					// range:  ---->|
+					lower = mid + 1
+				case -1:
+					// region: ---->|
+					// range:  ----->|
+					upper = mid - 1
+				}
+			}
+			uncachedRanges = uncachedRanges[mid:]
+			if bytes.Compare(endKey, uncachedRanges[0].StartKey) > 0 {
+				uncachedRanges[0].StartKey = endKey
+			}
+		}
+	}
+	// append the rest cache hit regions
+	for i := cachedIdx; i < len(cachedRegions); i++ {
+		appendRegion(cachedRegions[i])
+	}
```

Comment:
Better to wrap this merge operation into a standalone function and write unit-tests to cover all the possible combinations.